### PR TITLE
[IAM]: Custom policies support

### LINKED ISSUE #372: [IAM]: Implement custom policies management
Corresponding terraform issue: https://github.com/opentelekomcloud/terraform-provider-opentelekomcloud/issues/1767

### PR DESCRIPTION
**Pull request by artem-lifshits**
### What this PR does / why we need it
This PR implements management of custom policies for cloud servers/agencies
### Which issue this PR fixes
Closes: #372 

### Tests

=== RUN   TestPolicyLifecycle
--- PASS: TestPolicyLifecycle (3.57s)
=== RUN   TestAgencyPolicyLifecycle
--- PASS: TestAgencyPolicyLifecycle (4.48s)
=== RUN   TestList
--- PASS: TestList (0.84s)
PASS

Process finished with the exit code 0

### CODE REVIEW

File: acceptance/openstack/identity/v3/policies_test.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+	th.AssertEquals(t, newPolicy.DisplayName, allPages.Roles[0].DisplayName)
+	th.AssertEquals(t, newPolicy.Policy.Version, allPages.Roles[0].Policy.Version)
+	if !reflect.DeepEqual(newPolicy.Policy.Statement, allPages.Roles[0].Policy.Statement) {
+		t.Error("Statement parameters are different")
+	}
+
+	newName := tools.RandomString("policy-test-new-", 5)
+	newDescription := tools.RandomString("description-new-", 5)
+
+	bOpts := policies.CreateOpts{
```
Comment (anton-sidelnikov) [inline acceptance/openstack/identity/v3/policies_test.go]: aOpts and bOpts?

File: acceptance/openstack/identity/v3/policies_test.go [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+	}
+
+	client, err := clients.NewIdentityV3AdminClient()
+
+	th.AssertNoErr(t, err)
+
+	agencyOpts := agency.CreateOpts{
+		Name:            tools.RandomString("test-agency-", 5),
+		DomainID:        client.DomainID,
+		DelegatedDomain: "OTC-EU-DE-00000000001000020825",
```
Comment (anton-sidelnikov) [inline acceptance/openstack/identity/v3/policies_test.go]: keep domain name in env variable

File: openstack/identity/v3/agency/urls.go [hunk 3]
```diff
@@ -1,24 +1,32 @@
...
 
 const (
 	rootPath     = "OS-AGENCY"
 	resourcePath = "agencies"
 )
 
 func rootURL(c *golangsdk.ServiceClient) string {
-	return c.ServiceURL(rootPath, resourcePath)
+	url := c.ServiceURL(rootPath, resourcePath)
+	return strings.Replace(url, "/v3/", "/v3.0/", 1)
```
Comment (anton-sidelnikov) [inline openstack/identity/v3/agency/urls.go]: it didn't break anything?

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]

File: acceptance/openstack/identity/v3/policies_test.go (hunk 1 above)
Comment (artem-lifshits) [reply acceptance/openstack/identity/v3/policies_test.go]: First and second opts - one used for creation, another for resource update.
Should I rename them?

File: openstack/identity/v3/agency/urls.go (hunk 3 above)
Comment (artem-lifshits) [reply openstack/identity/v3/agency/urls.go]: Agencies weren't working to begin with, this url change fixed creation/deletion.
I needed that part for testing so I didn't check other functionality

File: acceptance/openstack/identity/v3/policies_test.go (hunk 1 above)
Comment (Aloento) [reply acceptance/openstack/identity/v3/policies_test.go]: use more descriptive names is better

File: acceptance/openstack/identity/v3/policies_test.go (hunk 2 above)
Comment (artem-lifshits) [reply acceptance/openstack/identity/v3/policies_test.go]: fixed

File: openstack/identity/v3/agency/urls.go (hunk 3 above)
Comment (artem-lifshits) [reply openstack/identity/v3/agency/urls.go]: I've left agency part as it was before and made some changes to client init, please check

File: acceptance/openstack/identity/v3/policies_test.go (hunk 1 above)
Comment (artem-lifshits) [reply acceptance/openstack/identity/v3/policies_test.go]: fixed

**Comment by artem-lifshits** [comment]
Pushed commit tests:

=== RUN   TestPolicyLifecycle
--- PASS: TestPolicyLifecycle (3.43s)
=== RUN   TestAgencyPolicyLifecycle
--- PASS: TestAgencyPolicyLifecycle (4.24s)
=== RUN   TestList
--- PASS: TestList (0.88s)
PASS

Process finished with the exit code 0

File: openstack/identity/v3/agency/urls.go (hunk 3 above)
Comment (Aloento) [reply openstack/identity/v3/agency/urls.go]: The endpoint of the IAM is `/v3.0/OS-${type}/${target}`
So it shouldn't break anything.
It is recommended that you make the replacement operation a method. The entire IAM needs it.

**Comment by artem-lifshits** [comment]
=== RUN   TestPolicyLifecycle
--- PASS: TestPolicyLifecycle (3.54s)
=== RUN   TestAgencyPolicyLifecycle
--- PASS: TestAgencyPolicyLifecycle (4.49s)
=== RUN   TestList
--- PASS: TestList (0.83s)
PASS

Process finished with the exit code 0

**Review by anton-sidelnikov — DISMISSED** [review]

File: openstack/identity/v3/policies/results.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+type CustomPolicy struct {
+	Version   string      `json:"Version"`
+	Statement []Statement `json:"Statement"`
+}
+
+type Statement struct {
+	Action    []string    `json:"Action"`
+	Effect    string      `json:"Effect"`
+	Condition Condition   `json:"Condition,omitempty"`
+	Resource  interface{} `json:"Resource,omitempty"`
```
Comment (Aloento) [inline openstack/identity/v3/policies/results.go:76]: Each type of Resource has a definite definition. Is it possible to consider making it specific?

**Review by vladimirvshivkov — DISMISSED** [review]

**Comment by Aloento** [comment]
Could you move the replace code to the function because not only the policies need it but also the whole IAM.

**Comment by artem-lifshits** [comment]
=== RUN   TestPolicyLifecycle
--- PASS: TestPolicyLifecycle (3.50s)
=== RUN   TestAgencyPolicyLifecycle
--- PASS: TestAgencyPolicyLifecycle (4.30s)
=== RUN   TestList
--- PASS: TestList (0.82s)
PASS

Process finished with the exit code 0

**Review by Aloento — APPROVED** [review]

**Pull request merged**